initrd/etc/luks-functions

```diff
@@ -267,24 +268,37 @@ prepare_thumb_drive()
 
 select_luks_container()
 {
+	#TODO: extend logic to prompt for block devices with model if multiple LUKS are found on block device instead of partitions
+	# Then feed luks with those partitions so that reencrypt and passphrase change can use passphrase to test all selected
 	TRACE_FUNC
 	if [ -s /boot/kexec_key_devices.txt ]; then
 		DEBUG "Reusing known good LUKS container device from /boot/kexec_key_devices.txt"
-		DEBUG "LUKS container device: $(cut -d ' ' -f1 /boot/kexec_key_devices.txt)"
-		LUKS=$(cut -d ' ' -f1 /boot/kexec_key_devices.txt)
-	else
+		LUKS=$(cut -d ' ' -f1 /boot/kexec_key_devices.txt) 
+		DEBUG "LUKS container device: $(echo $LUKS)"
+		# LUKS variable not exported yet, prompt for LUKS device
+	elif [ -z "$LUKS" ]; then
```

Comment:
I think this would be easier to understand if this function always attempted to prompt for a LUKS device.  Right now it's really difficult to see what is supposed to happen with the `LUKS` variable and I can't really tell if it'll work in various situations where you could cancel, try again, make a mistake, etc.

- `luks_reencrypt` seems to fight with this function trying to reuse `LUKS`.  This one seems to reuse it if it's already set (I guess?), but `luks_reencrypt` doesn't want that so it unsets the variable in some situation, and to be frank I can't tell where it actually loops (does it?) that any of that could matter.
- `test_luks_current_disk_recovery_key_passphrase` which does loop, but doesn't know to unset LUKS, but does also export LUKS for some reason (not sure why).
- `luks_change_passphrase` doesn't loop around the selection, but does have an unset and an export.

If this function always attempted to prompt for a LUKS device, we could make it clear where LUKS might already be set and not prompt, or be sure we are prompting again if we call it again.  We could also make it clear where we might take an existing value from /boot.

If possible I'd encourage printing the selected value to stdout and capturing it in the caller (fewer global variables makes data flow much easier to understand), but sometimes that is tricking when prompting.  (It's possible though!)  If the prompting makes that too difficult, it'd be OK to store it in a variable, but always do the prompt and store it, let the callers handle business logic to decide whether to prompt I think.

It'd also be worthwhile to document what this function is trying to do, which I can't really isolate right now since it's pretty glued to the functions calling it.

Reply:
> I think this would be easier to understand if this function always attempted to prompt for a LUKS device. Right now it's really difficult to see what is supposed to happen with the LUKS variable and I can't really tell if it'll work in various situations where you could cancel, try again, make a mistake, etc.

As seen under https://github.com/linuxboot/heads/pull/1541/files#diff-2bca82e612385576acc67c4a36b2a9cf4d563281f5aa115b2b58fabf68e789e8R276

if $LUKS filled by configured TPM DUK provisioned (which is the only place filling /boot/kexec_key_devices.txt), it is reused.
 

> luks_reencrypt seems to fight with this function trying to reuse LUKS. This one seems to reuse it if it's already set (I guess?), but luks_reencrypt doesn't want that so it unsets the variable in some situation, and to be frank I can't tell where it actually loops (does it?) that any of that could matter.

luks_reencrypt, test_luks_current_disk_recovery_key_passphrase and luks_change_passphrase will unset $LUKS if ops don't work, recalling slect_luks_container and test_luks_current_disk_recovery_key_passphrase. I tested pretty thoroughly luks_reenctypt and  luks_change_passphrase codepaths direcyl without issue (menu options), as well as re-ownership
and TPM DUK when setting boot default option optional LUKS TPM DUK, which fails if two containers not sharing same DUK, which also gives warning for user to fix manually through Option for LUKS passphrase change.

Reply:
> It'd also be worthwhile to document what this function is trying to do, which I can't really isolate right now since it's pretty glued to the functions calling it.

Will do

Reply:
@JonathonHall-Purism this thread should answer your initial comment at https://github.com/linuxboot/heads/pull/1541#discussion_r1735115819